src/System.Management.Automation/engine/LanguagePrimitives.cs

```diff
@@ -4772,66 +4787,138 @@ internal static object ThrowInvalidConversionException(object valueToConvert, Ty
                 return CacheConversion<object>(fromType, toType, LanguagePrimitives.ConvertIntegerToEnum, ConversionRank.Language);
             }
 
-            if (fromType.IsSubclassOf(typeof(PSMethod)) && toType.IsSubclassOf(typeof(Delegate)))
+            if (fromType.IsSubclassOf(typeof(PSMethod)) && toType.IsSubclassOf(typeof(Delegate)) && !toType.IsAbstract)
             {
-                var mi = toType.GetMethod("Invoke");
-
-                var comparator = new DelegateArgsComparator(mi);
+                var targetMethod = toType.GetMethod("Invoke");
+                var comparator = new SignatureComparator(targetMethod);
                 var signatureEnumerator = new PSMethodSignatureEnumerator(fromType);
+                int index = -1, matchedIndex = -1;
+
                 while (signatureEnumerator.MoveNext())
                 {
-                    var candidate = signatureEnumerator.Current.GetMethod("Invoke");
-                    if (comparator.SignatureMatches(candidate.ReturnType, candidate.GetParameters()))
+                    index ++;
+                    var signatureType = signatureEnumerator.Current;
+                    // Skip the non-bindable signatures
+                    if (signatureType == typeof(Func<PSNonBindableType>)) { continue; }
+
+                    Type[] argumentTypes = signatureType.GenericTypeArguments;
+                    if (comparator.ProjectedSignatureMatchesTarget(argumentTypes, out bool signaturesMatchExactly))
                     {
-                        return CacheConversion<Delegate>(fromType, toType, LanguagePrimitives.ConvertPSMethodInfoToDelegate, ConversionRank.Language);
+                        if (signaturesMatchExactly)
+                        {
+                            // We prefer the signature that exactly matches the target delegate.
+                            matchedIndex = index;
+                            break;
+                        }
+
+                        // If there is no exact match, then we use the compatible that was first found.
```

Comment:
> // If there is no exact match, then we use the first compatible signature we found.
   if (matchedIndex == -1) { matchedIndex = index; }

In case no exact match is found, It sticks to the first compatible one. When you find the first compatible one, `matchedIndex` will be set to the index and thus it's not -1 anymore, so subsequent compatible matches won't change its value.

There are definitely ways to do better than the first compatible match in case there is no exact match, but I feel that's much complexities with no much add-on value for this feature, so I choose to not consider a more sophisticated resolution.

Reply:
Ah, right, yes I see now! I was thinking it through, but just wasn't quite getting there — figured it made sense to ask you how it works.

I definitely agree that we don't want to expand the PR to implement a more complicated version. My only thought is that now we have something implemented, changing it later becomes a breaking change rather than a simple feature addition/improvement.

Reply:
> changing it later becomes a breaking change rather than a simple feature addition/improvement.

Good point. That could be a problem. I think if it turns out we really need to implement a more complex resolution approach, probably there is a scenario strong enough to justify the breaking change (considering the existing behavior is a must-fix bug).